File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/til_sed/lovvalg/A010Mapper.java
package no.nav.melosys.eessi.service.sed.mapper.til_sed.lovvalg;

import java.util.Optional;
import java.util.Set;

import no.nav.melosys.eessi.controller.dto.Bestemmelse;
import no.nav.melosys.eessi.controller.dto.Lovvalgsperiode;
import no.nav.melosys.eessi.controller.dto.SedDataDto;
import no.nav.melosys.eessi.models.*;
import no.nav.melosys.eessi.models.exception.MappingException;
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA010;
import no.nav.melosys.eessi.models.sed.nav.*;

import static no.nav.melosys.eessi.controller.dto.Bestemmelse.*;

public class A010Mapper implements LovvalgSedMapper<MedlemskapA010> {

    private static final Set<Bestemmelse> LOVLIGE_BESTEMMELSER_A010 = Set.of(ART_11_3_b, ART_11_3_c, ART_11_3_d, ART_11_4, ART_11_5, ART_15);

    @Override
    public MedlemskapA010 getMedlemskap(SedDataDto sedData) {
        MedlemskapA010 medlemskap = new MedlemskapA010();
        Optional<Lovvalgsperiode> lovvalgsperiode = sedData.finnLovvalgsperiode();

        lovvalgsperiode.ifPresent(value -> medlemskap.setMeldingomlovvalg(hentMeldingOmLovvalg(value)));

        medlemskap.setVedtak(hentVedtak(sedData));
        medlemskap.setAndreland(getAndreland(sedData));
        return medlemskap;
    }

    private VedtakA010 hentVedtak(SedDataDto sedDataDto){
        VedtakA010 vedtak = new VedtakA010();
        final Optional<Lovvalgsperiode> lovvalgsperiode = sedDataDto.finnLovvalgsperiode();
        if (lovvalgsperiode.isPresent()){
            vedtak.setGjelderperiode(hentPeriode(lovvalgsperiode.get()));
            vedtak.setLand(lovvalgsperiode.get().getLovvalgsland());
        }

        setVedtaksdata(vedtak, sedDataDto.getVedtakDto());
        vedtak.setGjeldervarighetyrkesaktivitet("ja");
        return vedtak;
    }

    private Utsendingsland getAndreland(SedDataDto sedData) {
        final String lovvalgsland = sedData.finnLovvalgslandDefaultNO();
        Utsendingsland utsendingsland = new Utsendingsland();
        utsendingsland.setArbeidsgiver(hentArbeidsgivereIkkeILand(sedData.getArbeidsgivendeVirksomheter(), lovvalgsland));
        return utsendingsland;
    }

    private MeldingOmLovvalg hentMeldingOmLovvalg(Lovvalgsperiode lovvalgsperiode) {
        MeldingOmLovvalg meldingOmLovvalg = new MeldingOmLovvalg();
        meldingOmLovvalg.setArtikkel(tilA010Bestemmelse(lovvalgsperiode));
        return  meldingOmLovvalg;
    }

    private String tilA010Bestemmelse(Lovvalgsperiode lovvalgsperiode) {

        if (LOVLIGE_BESTEMMELSER_A010.contains(lovvalgsperiode.getBestemmelse())) {
            return lovvalgsperiode.getBestemmelse().getValue();
        } else if (lovvalgsperiode.harTilleggsbestemmelse() && LOVLIGE_BESTEMMELSER_A010.contains(lovvalgsperiode.getTilleggsBestemmelse())) {
            return lovvalgsperiode.getTilleggsBestemmelse().getValue();
        }

        throw new MappingException("Kan ikke mappe til bestemmelse i A010 for lovvalgsperiode {}");
    }

    private PeriodeA010 hentPeriode(Lovvalgsperiode lovvalgsperiode) {
        PeriodeA010 periode = new PeriodeA010();
        periode.setStartdato(formaterDato(lovvalgsperiode.getFom()));
        periode.setSluttdato(formaterDato(lovvalgsperiode.getTom()));
        return periode;
    }

    @Override
    public SedType getSedType() {
        return SedType.A010;
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/til_sed/lovvalg/A010Mapper.kt
package no.nav.melosys.eessi.service.sed.mapper.til_sed.lovvalg

import no.nav.melosys.eessi.controller.dto.Bestemmelse
import no.nav.melosys.eessi.controller.dto.Lovvalgsperiode
import no.nav.melosys.eessi.controller.dto.SedDataDto
import no.nav.melosys.eessi.models.SedType
import no.nav.melosys.eessi.models.exception.MappingException
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA010
import no.nav.melosys.eessi.models.sed.nav.MeldingOmLovvalg
import no.nav.melosys.eessi.models.sed.nav.PeriodeA010
import no.nav.melosys.eessi.models.sed.nav.Utsendingsland
import no.nav.melosys.eessi.models.sed.nav.VedtakA010

class A010Mapper : LovvalgSedMapper<MedlemskapA010> {

    override fun getSedType() = SedType.A010

    override fun getMedlemskap(sedData: SedDataDto): MedlemskapA010 {
        val lovvalgsland = sedData.finnLovvalgslandDefaultNO()
        val lovvalgsperiode = sedData.finnLovvalgsperiode()

        return MedlemskapA010(
            meldingomlovvalg = sedData.finnLovvalgsperiode()?.let {
                MeldingOmLovvalg(artikkel = tilA010Bestemmelse(it))
            },
            vedtak = VedtakA010(
                gjelderperiode = lovvalgsperiode?.let { hentPeriode(it) },
                land = lovvalgsperiode?.lovvalgsland,
                gjeldervarighetyrkesaktivitet = "ja"
            ).also {
                setVedtaksdata(it, sedData.vedtakDto)
            },
            andreland = Utsendingsland(
                arbeidsgiver = hentArbeidsgivereIkkeILand(sedData.arbeidsgivendeVirksomheter, lovvalgsland)
            )
        )
    }

    private fun tilA010Bestemmelse(lovvalgsperiode: Lovvalgsperiode): String {
        val bestemmelse = lovvalgsperiode.bestemmelse ?: throw IllegalArgumentException("lovvalgsperiode.bestemmelse kan ikke være null")

        return when {
            LOVLIGE_BESTEMMELSER_A010.contains(bestemmelse) ->
                bestemmelse.value

            lovvalgsperiode.harTilleggsbestemmelse() &&
                LOVLIGE_BESTEMMELSER_A010.contains(lovvalgsperiode.tilleggsBestemmelse) ->
                lovvalgsperiode.tilleggsBestemmelse!!.value

            else -> throw MappingException("Kan ikke mappe til bestemmelse i A010 for lovvalgsperiode $lovvalgsperiode")
        }
    }

    private fun hentPeriode(lovvalgsperiode: Lovvalgsperiode) = PeriodeA010(
        startdato = lovvalgsperiode.fom.formater(),
        sluttdato = lovvalgsperiode.tom.formater()
    )

    companion object {
        private val LOVLIGE_BESTEMMELSER_A010 = setOf(
            Bestemmelse.ART_11_3_b,
            Bestemmelse.ART_11_3_c,
            Bestemmelse.ART_11_3_d,
            Bestemmelse.ART_11_4,
            Bestemmelse.ART_11_5,
            Bestemmelse.ART_15
        )
    }
}
